academy/lending-protocol/Implementation Guide.md

```diff
@@ -0,0 +1,254 @@
+### Complete Implementation Guide: Lending Protocol on =nil; Foundation
+
+This guide provides a **step-by-step walkthrough** for the example implementation of the decentralized lending and borrowing protocol on **=nil;**, focusing on **sharded smart contract architecture** and the lifecycle of core actions like **opening a loan**, **borrowing funds**, and **repaying the loan**. We'll go through each contract, their functions, how they interact, and the details of important function calls, including the **`sendRequestWithTokens`, `asyncCall`** methods and how they are used throughout the system.
+
+### 1. **Deployment and Contract Overview**
+
+Before diving into the workflows for opening a loan, borrowing funds, and repaying a loan, it’s important to understand the smart contracts and their deployment.
+
+#### **Core Contracts**
+
+1. **GlobalLedger**: This contract tracks user deposits and loans, serving as the centralized state manager for the protocol. It stores the user’s deposit balance and the loan balance.
+2. **InterestManager**: This contract manages interest rates, which are used when calculating the repayment amounts for loans.
+3. **LendingPool**: This is the core contract for interacting with the lending protocol. It handles user deposits, loans, and repayments.
+4. **Oracle**: Provides the token price data needed for collateral calculations when a loan is opened.
+
+The deployment process occurs using Hardhat, where the smart contracts are deployed on different shards:
+
+- **InterestManager** (Shard 2)
+- **GlobalLedger** (Shard 3)
+- **Oracle** (Shard 4)
+- **LendingPool** (Shard 1)
+
```

Comment:
we should be clear that we deploy on different shards to illustrate the programming patterns you can implement with =nil;. App design recommendations are case-specific and out-of-scope here anyway